Sources/Jobs/JobQueueHandler.swift

```diff
@@ -94,34 +96,44 @@ final class JobQueueHandler<Queue: JobQueueDriver>: Sendable {
 
         do {
-            while true {
-                do {
-                    Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
-                    try await job.execute(context: .init(logger: logger))
-                    break
-                } catch let error as CancellationError {
-                    logger.debug("Job cancelled")
-                    // Job failed is called but due to the fact the task is cancelled, depending on the
-                    // job queue driver, the process of failing the job might not occur because itself
-                    // might get cancelled
+            do {
+                Meter(label: self.meterLabel, dimensions: [("status", JobStatus.processing.rawValue)]).increment()
+                try await job.execute(context: .init(logger: logger))
+            } catch let error as CancellationError {
+                logger.debug("Job cancelled")
+                // Job failed is called but due to the fact the task is cancelled, depending on the
+                // job queue driver, the process of failing the job might not occur because itself
+                // might get cancelled
+                try await self.queue.failed(jobId: queuedJob.id, error: error)
+                self.updateJobMetrics(for: job.name, startTime: startTime, error: error)
+                return
+            } catch {
+                if job.didFail {
+                    logger.debug("Job: failed")
                     try await self.queue.failed(jobId: queuedJob.id, error: error)
                     self.updateJobMetrics(for: job.name, startTime: startTime, error: error)
                     return
-                } catch {
-                    if count <= 0 {
-                        logger.debug("Job failed")
-                        try await self.queue.failed(jobId: queuedJob.id, error: error)
-                        self.updateJobMetrics(for: job.name, startTime: startTime, error: error)
-                        return
-                    }
-                    count -= 1
-                    logger.debug("Retrying Job")
-                    self.updateJobMetrics(for: job.name, startTime: startTime, retrying: true)
                 }
+
+                let attempts = (job.attempts ?? 0) + 1
+
+                let delay = self.backoff(attempts: attempts)
+
+                // remove from processing lists
+                try await self.queue.finished(jobId: queuedJob.id)
+                // push new job in the queue
+                _ = try await self.queue.push(
+                    self.queue.encode(job, attempts: attempts),
+                    options: .init(
+                        delayUntil: delay
+                    )
+                )
+                self.updateJobMetrics(for: job.name, startTime: startTime, retrying: true)
+                logger.debug("Retrying Job with attempts: \(attempts) and delayed until: \(delay)")
```

Comment:
Logger messages should really be static and any dynamic elements be metadata eg
```suggestion
                logger.debug(
                    "Retrying Job", 
                    metadata: [
                        "attempts": .stringConvertible(attempts), 
                        "delayedUntil": .stringConvertible(delay)
                    ]
                )
```